include/picongpu/plugins/transitionRadiation/Calculator.hpp

```diff
@@ -0,0 +1,238 @@
+/* Copyright 2013-2019 Heiko Burau, Rene Widera, Richard Pausch, Finn-Ole Carstens
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "Particle.hpp"
+
+
+namespace picongpu
+{
+namespace transitionRadiation
+{
+    using complex_X = pmacc::math::Complex< float_X >;
+    using complex_64 = pmacc::math::Complex< float_64 >;
+
+    /* Arbitrary margin which is necessary to prevent division by 0 error
+        * created by particles moving in the plane of the foil.
+        */
+    float_X const DIV_BY_ZERO_MINIMUM = 1.e-7;
+
+    /** Calculator class for calculation of transition radiation.
+     * 
+     * @param particleSet transitionRadiation::Particle to compute transition radiation for
+     * @param lookDirection vector with observation direction
+     */
+    class Calculator
+    {
+
+    private:
+        transitionRadiation::Particle const & particle;
+        float3_X const lookDirection;
+
+        float_X parMomSinTheta;
+        float_X parMomCosTheta;
+        float_X const parMomPhi;
+        float_X parMomSinPhi;
+        float_X parMomCosPhi;
+        float_X detectorSinTheta;
+        float_X detectorCosTheta;
+        float_X const detectorPhi;
+        float_X const parSqrtOnePlusUSquared;
+
+    public: 
+        HDINLINE
+        Calculator(
+            transitionRadiation::Particle const & particleSet,
+            float3_X const lookDirection
+        ) : 
+            particle( particleSet ),
+            lookDirection( lookDirection ),
+            parMomPhi( particle.getMomPhi( ) ),
+            detectorPhi( 
+                picongpu::math::atan2(
+                    lookDirection.z( ), 
+                    lookDirection.x( )
+                ) + picongpu::PI
+            ),
+            parSqrtOnePlusUSquared( 
+                picongpu::math::sqrt( 1 + particle.getU( ) * particle.getU( ) ) 
```

Comment:
`particle.getU( ) * particle.getU( )` is reused in line 119 - you could save the result in a (register) variable and thus avoid recomputation (5 ALU operations - possible only 3 due to `*+` optimization) 

@psychocoderHPC what are your thoughts on such optimizations?